plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/AbstractOpenshiftUIElement.java

```diff
@@ -0,0 +1,63 @@
+package org.jboss.tools.openshift.internal.ui.models;
+
+import org.eclipse.core.runtime.IAdaptable;
+
+public abstract class AbstractOpenshiftUIElement<R, P extends IOpenshiftUIElement<?>> implements IAdaptable, IOpenshiftUIElement<P> {
+	private P parent;
+	private R wrapped;
+
+	public AbstractOpenshiftUIElement(P parent, R wrapped) {
+		this.parent = parent;
+		this.wrapped= wrapped;
+	}
+
+	public P getParent() {
+		return parent;
+	}
+	
+	public R getWrapped() {
+		return wrapped;
+	}
+	
+	@SuppressWarnings("unchecked")
+	public synchronized <T> T getAdapter(Class<T> adapter) {
+		if (adapter.isInstance(this)) {
+			return (T) this;
+		} else if (adapter.isInstance(wrapped)) {
+			return (T) wrapped;
+		}
+		return null;
+	}
+	
+	@Override
+	public boolean equals(Object o) {
+		if (o == null || o.getClass() != getClass()) {
+			return false;
+		}
+		if (o == this) {
+			return true;
+		}
+		AbstractOpenshiftUIElement<?, ?> other= (AbstractOpenshiftUIElement<?, ?>) o;
+		return wrapped.equals(other.getWrapped()) && getParent().equals(other.getParent());
+	}
+	
+	@Override
+	public int hashCode() {
+		return wrapped.hashCode();

```

Comment:
Since only few elements will have the same wrapped element but a different parent, this will add little to the quality of the hash function.
Note that returning 42 would be a very bad, but perfectly legal hash function.